File: python/assn1-tdd-car/car.py
import time

class Car:
    OFF = 0
    ON = 1
    FULL = 1
    EMPTY = 0
    MAXFUEL = 13
    def __init__(self):
        """car has 4 wheels and standard weight of 1500lbs with 200hp, not like that's important"""
        self._enginestate = self.OFF
        self._speed = 0  # MPH
        self._tankstate = self.EMPTY
        self._fuellevel = 0  # Gallons

    def startEngine(self):
        if self._enginestate == self.OFF:
            self._enginestate = self.ON
        else:
            raise Exception("Engine is already on")

    def stopEngine(self):
        if self._speed == 0 and self._enginestate == self.ON:
            self._enginestate = self.OFF
        else:
            raise Exception("Engine is already off")

    def fillTank(self, amount):
        if amount + self._fuellevel <= self.MAXFUEL:
            self._fuellevel += amount
        else:
            raise Exception("Cant fill tank without overflowing")

    def _engineTick(self):  # 1 tick = 1 RPM
        if self._enginestate == self.ON:
            if self._fuellevel > 0:
                self._fuellevel -= 0.0001
                print("\r Fuel level: {} gal".format(self._fuellevel), end="")
            else:
                self.stopEngine()

    def runUntilEmpty(self, rpm):  # Run engine until tank is gone at rpm
        rps = rpm // 60
        while self._fuellevel > 0:
            self._engineTick()
            time.sleep(1/rps)

if __name__ == "__main__":
    car = Car()
    car.fillTank(13)
    car.startEngine()
    print("Engine Started")
    car.runUntilEmpty(1500000)

File: java/car/src/net/jtcressy/csci375/Car.java

package net.jtcressy.csci375;

public class Car {
    public enum CarState {
        ON, OFF, FULL, EMPTY
    }
    private static float MAXFUEL = 13;
    private CarState carstate;
    private CarState tankstate;
    private int speed;
    private float fuellevel;
    Car() {
        carstate = CarState.OFF;
        tankstate = CarState.EMPTY;
        speed = 0;
        fuellevel = 0;
    }

    public void startEngine() {
        if (carstate == CarState.OFF) {
            carstate = CarState.ON;
            try {
                runUntilEmpty(1500);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public void stopEngine() {
        if (speed > 0 && tankstate != CarState.EMPTY) {
            //cant turn off car in motion, but if the tank is empty it has to stop anyway.
            throw new IllegalStateException("Car is in motion, cannot stop engine.");
        }
        if (carstate == CarState.OFF) {
            //its already off, duh.
            throw new IllegalStateException("Car is already off, cannot turn off");
        }
        carstate = CarState.OFF;
    }

    public void fillTank(int amount) {
        if (amount + fuellevel > MAXFUEL) {
            //cant fill a tank beyond maximum
            throw new IllegalArgumentException("Tank is either full or cannot be filled past its maximum capacity.");
        }
        fuellevel += amount;
    }

    private void engineTick() {
        if (carstate == CarState.ON) {
            if (fuellevel > 0.01) {
                fuellevel -= 0.01;
            }
            else {
                fuellevel = 0; //Prevents fuel from going negative
            }
            String lineout = String.format("\rFuel Level: %s gal", fuellevel);
            System.out.print(lineout);
        }
    }

    public void runUntilEmpty(int rpm) throws InterruptedException {
        long rps = rpm / 60;
        while (fuellevel > 0) {
            engineTick();
            Thread.sleep(1000 / rps);
        }
    }
}
